src/IrcTransport/IrcTcpTransport.cs

```diff
@@ -0,0 +1,523 @@
+        /// </summary>
+        public event ReadLineEventHandler OnMessageReceived;
+
+        /// <summary>
+        /// Event which fires when a connection error occurs
+        /// </summary>
+        public event Action OnConnectionError;
+
+        /// <summary>
+        /// Timeout in seconds for receiving data from the socket
+        /// Default: 600
+        /// </summary>
+        public int SocketReceiveTimeout { get; set; } = 600;
+
+        /// <summary>
+        /// Timeout in seconds for sending data to the socket
+        /// Default: 600
+        /// </summary>
+        public int SocketSendTimeout { get; set; } = 600;
+
+        /// <summary>
+        /// Connection host
+        /// </summary>
+        public string Address { get; set; }
+
+        /// <summary>
+        /// Connection port
+        /// </summary>
+        public int Port { get; set; }
+
+        /// <summary>
+        /// If you want to use a Proxy, set the ProxyHost to Host of the Proxy you want to use.
+        /// </summary>
+        public string ProxyHost { get; set; }
+
+        /// <summary>
+        /// If you want to use a Proxy, set the ProxyPort to Port of the Proxy you want to use.
+        /// </summary>
+        public int ProxyPort { get; set; }
+
+        /// <summary>
+        /// Standard Setting is to use no Proxy Server, if you Set this to any other value,
+        /// you have to set the ProxyHost and ProxyPort aswell (and give credentials if needed)
+        /// Default: ProxyType.None
+        /// </summary>
+        public ProxyType ProxyType { get; set; } = ProxyType.None;
+
+        /// <summary>
+        /// Username to your Proxy Server
+        /// </summary>
+        public string ProxyUsername { get; set; }
+
+        /// <summary>
+        /// Password to your Proxy Server
+        /// </summary>
+        public string ProxyPassword { get; set; }
+
+        /// <summary>
+        /// Enables/disables using SSL for the connection
+        /// Default: false
+        /// </summary>
+        public bool UseSsl { get; set; }
+
+        /// <summary>
+        /// Specifies if the certificate of the server is validated
+        /// Default: true
+        /// </summary>
+        public bool ValidateServerCertificate { get; set; }
+
+        /// <summary>
+        /// Specifies the client certificate used for the SSL connection
+        /// Default: null
+        /// </summary>
+        public X509Certificate SslClientCertificate { get; set; }
+
+        /// <summary>
+        /// The encoding to use to write to and read from the socket.
+        ///
+        /// If EnableUTF8Recode is true, reading and writing will always happen
+        /// using UTF-8; this encoding is only used to decode incoming messages
+        /// that cannot be successfully decoded using UTF-8.
+        ///
+        /// Default: encoding of the system
+        /// </summary>
+        public Encoding Encoding { get; set; } = Encoding.Default;
```

Comment:
Besides, I believe C#6 is not included in .NET4.0, maybe we're talking about .NET4.5 or higher.